File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool res = true;
    bool isBalanced(TreeNode* root) {
        traverse(root);
        return res;
    }

    int traverse(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }

        int left = traverse(root->left);

        int right = traverse(root->right);

        if (abs(left - right) > 1) {
            res = false;
        }

        return max(left, right) + 1;
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True 

        def findDepth(curr):
            if not curr:
                return 0
            
            leftDepth = findDepth(curr.left)
            rightDepth = findDepth(curr.right)

            if leftDepth == -1 or rightDepth == -1:
                return -1
            elif abs(leftDepth - rightDepth) > 1:
                return -1
            else:
                return max(leftDepth, rightDepth) + 1

        return findDepth(root) != -1
            